src/components/responsive-flyout/ResponsiveFlyout.js

```diff
@@ -0,0 +1,70 @@
+// @flow
+import * as React from 'react';
+import classNames from 'classnames';
+import { Flyout, Overlay } from '../flyout';
+import ScrollWrapper from '../scroll-wrapper';
+import type { FlyoutProps } from '../flyout/Flyout';
+import OverlayHeader from './OverlayHeader';
+
+import './styles/ResponsiveFlyout.scss';
+
+type Props = FlyoutProps & {
+    /** Components to render in the overlay */
+    children: ?React.Node,
+    /** Set className to the overlay wrapper */
+    className?: string,
+    /** Disable responsiveness behavior for higher-level components that want control */
+    enabled?: boolean,
+    /** Custom button to trigger for opening/closing the flyout */
+    flyoutButton: React.Element<any>,
+    /** Optional function to get the flyoutRef in parent components */
+    flyoutRefFn?: any => any,
+    /** Header for responsive layout */
+    header?: React.Element<any>,
+    /** Overlay classname */
+    overlayClassName?: string,
+    /** Optional function to get the scrollRef in parent components */
+    scrollRefFn?: any => any,
+};
+
+class ResponsiveFlyout extends React.Component<Props> {
+    static defaultProps = {
+        position: 'bottom-left',
+        enabled: true,
+    };
+
+    constructor(props) {
+        super(props);
+        this.flyoutRef = React.createRef();
+    }
+
+    render() {
+        const {
+            header,
+            flyoutButton,
+            children,
+            enabled,
+            scrollRefFn,
+            className,
+            overlayClassName,
+            flyoutRefFn,
+            ...rest
+        } = this.props;
+
+        return (
+            <Flyout ref={this.flyoutRef} className={classNames({ 'responsive-flyout': enabled }, className)} {...rest}>
+                {flyoutButton}
+                <Overlay className={overlayClassName}>
+                    <OverlayHeader onCloseClick={() => this.flyoutRef.current.closeOverlay()}>{header}</OverlayHeader>
```

Comment:
I opted to use refs because I felt that it would be cleaner to implement without needing to make changes to Flyout or Overlay

I looked at passing it down via Context Providers, but I didn't like that solution either and seemed like it was an anti-pattern according to React. Using refs in this way also seemed like an anti-pattern as well, but was the lesser of two evils.

The "more correct" solution might be to create a ResponsiveOverlay and implement OverlayHeader in there and replace Overlay here with it. ResponsiveOverlay would take in a header prop. This way we can handle it declaratively and avoid using refs

Reply:
Oh, I just saw this. Great thinking, for sure!